Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+            
+            let padding: UInt8
+            let dataLen: Int
+            if flags.contains(.padded) {
+                padding = bytes.readInteger()!
+                dataLen = header.length - 1 - Int(padding)
+                if dataLen <= 0 {
+                    // If the length of the padding is the length of the frame payload or greater,
+                    // the recipient MUST treat this as a connection error (Section 5.4.1) of type
+                    // PROTOCOL_ERROR.
+                    throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+                }
+            } else {
+                padding = 0
+                dataLen = header.length
+            }
+            
+            let buf: ByteBuffer
+            if bytes.readableBytes == dataLen {
+                buf = bytes
+            } else {
+                buf = bytes.readSlice(length: dataLen)!
+            }
+            
+            payload = .data(.byteBuffer(buf))
+            
+        case 1:
+            // HEADERS frame : RFC 7540 § 6.2
+            guard streamID != .rootStream else {
+                // HEADERS frames MUST be associated with a stream. If a HEADERS frame is received whose
+                // stream identifier field is 0x0, the recipient MUST respond with a connection error
+                // (Section 5.4.1) of type PROTOCOL_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+            }
+            
+            let stream = HTTP2StreamData.getStream(for: streamID, allocator: self.allocator)
+            var bytesToRead = header.length
+            
+            let padding: UInt8
+            if flags.contains(.padded) {
+                padding = bytes.readInteger()!
+                bytesToRead -= 1
+                if bytesToRead <= Int(padding) {
+                    // Padding that exceeds the size remaining for the header block fragment
+                    // MUST be treated as a PROTOCOL_ERROR.
+                    throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+                }
+            } else {
+                padding = 0
+            }
+            
+            let priorityData: HTTP2Frame.StreamPriorityData?
+            if flags.contains(.priority) {
+                let raw: UInt32 = bytes.readInteger()!
+                priorityData = HTTP2Frame.StreamPriorityData(exclusive: (raw & 0x8000_0000 != 0),
+                                                             dependency: HTTP2StreamID(knownID: Int32(raw & ~0x8000_0000)),
+                                                             weight: bytes.readInteger()!)
+                bytesToRead -= 5
+            } else {
+                priorityData = nil
+            }
+            
+            // slice the input buffer if necessary
+            let headerByteSize = bytesToRead - Int(padding)
+            let headers: HPACKHeaders
+            if bytes.readableBytes == headerByteSize {
```

Comment:
It may be easier in all of these blocks just to perform the unconditional slice operation, rather than branch on the length. This branch will so rarely be taken that it's almost certainly not going to help.

Reply:
(This note applies to all of these blocks).